agent/qrexec-agent-data.c

```diff
@@ -500,6 +509,209 @@ static int process_child_io(libvchan_t *data_vchan,
+                set_nonblock(s);
+                return 0;
+            }
+            switch (errno) {
+            // These cannot happen
+            case EFAULT:       // all of our parameters are in valid memory
+            case EINVAL:       // we passed valid parameters
+            case ENOTSOCK:     // ditto
+            case ENETUNREACH:  // cannot happen for AF_UNIX
+            case EADDRNOTAVAIL:// cannot happen for AF_UNIX
+            case EINPROGRESS:  // this socket is blocking
+            case EBADF:        // `s` was created by a call to `socket()`
+            case EAFNOSUPPORT: // the kernel supports AF_UNIX
+            case EALREADY:     // ditto
+                abort();
+            // These should not happen
+            case EAGAIN:       // ditto
+#if EAGAIN != EWOULDBLOCK
+            case EWOULDBLOCK:  // ditto
+#endif
+                // Probably a misbehaving FUSE filesystem
+                break;
+            case EINTR:
+                // Interrupted by a signal - retry
+                --i; // to retry loop iteration
+                continue;
+            case ENOENT:
+            case EISDIR:
+            case ENOTDIR:
+            case EIO:
+                // These errors could also happen with `execve()`
+                break;
+            case EPROTOTYPE:
+            case ETIMEDOUT:
+                // Socket server problem.  Fail the whole connection.
+                // (We do not want to fall back because the user may have overriden
+                // behavior for security reasons, and because fail-fast is much
+                // easier to debug).
+                goto fail;
+            case EACCES:
+            case EPERM:
+            case ECONNREFUSED: {
+	        if (access(remote.sun_path, X_OK) != 0) {
+                    fprintf(stderr, "cannot execute %s: no execute permission\n", remote.sun_path);
+                    break;
+                }
+                struct stat stat_buf;
+                if (stat(remote.sun_path, &stat_buf) != 0) {
+                    fprintf(stderr, "cannot execute %s: cannot stat\n", remote.sun_path);
+                    break;
+                }
+                if (!S_ISREG(stat_buf.st_mode)) {
+                    fprintf(stderr, "cannot execute %s: not a regular file\n", remote.sun_path);
+                    break;
+                }
+                if ((stat_buf.st_mode & 022)) {
+                    fprintf(stderr, "refusing to execute %s: group- or world- writable\n", remote.sun_path);
+                    break;
+                }
+                if ((stat_buf.st_uid && stat_buf.st_uid != getuid())) {
+                    fprintf(stderr, "refusing to execute %s: neither owned by root or myself\n", remote.sun_path);
+                    break;
+                }
+                // Indicates that this is a regular file
+                char *const arguments[4] __attribute__((unused)) = {
```

Comment:
Why? If preparation for replacement of `qubes-rpc-multiplexer`, better introduce it in separate commit, as this is unrelated to socket-based services.

Reply:
Agreed.